### PR TITLE
Godocker job runner

### PR DESCRIPTION
**Pull request by dannon**
This reverts the commits in #2700 which stripped out the GoDocker job runner from #2653.

Essentially, treat this as a re-open of #2653, I just had to jump through a hoop or two to get here.

**Google Summer of Code 2016: Open genome informatics
_Project Idea 13: Using Go-Docker to enhance Galaxy functionalities_**

A REST based runner for executing Galaxy jobs using GoDocker.

Go-Docker is a cluster management tool using Docker as execution/isolation system. It can be seen like Sun Grid Engine/Torque/… Using Bioinformatics dedicated Docker hub as BioShaDock, it allows Life scientists to use powerful tools in a easy and secure way while finely managing infrastructures resources. Galaxy is a user-friendly web platform originally dedicated to life sciences data analysis. If Galaxy provides amazing functionalities to enhance transparency, accessibility and reproducibility of data analysis related processes, there is still opportunities to provide more reproducibility and flexibility in the manner of using this web platform. This project idea is to integrate the Go-Docker system with Galaxy.

Galaxy is designed to run jobs on your local system by default, but it can be configured to run jobs on a cluster. The front-end Galaxy application runs on a single server as usual, but tools are run on cluster nodes instead. Here, we use GoDocker system to manage authentication, scheduling and execution of jobs.

### CODE REVIEW

**Comment by dannon** [comment]
I'm overhauling some of the logging in a followup commit.

**Comment by varunshankar** [comment]
Thanks for the changes @dannon. As galaxy has many classes, the get_structure method in runner code can be used to easily find out the class of an object.

**Comment by osallou** [comment]
@dannon , could you test the runner in the vagrant box ? what is merge status?
Thanks

**Comment by osallou** [comment]
@dannon, any news on merge ? Seems tests were ok?
Thanks

**Comment by dannon** [comment]
@osallou I did get it to work in the box, but because of how I had to reopen this PR I cannot be the one to merge.  I just realized this was still tagged WIP, let me request review from someone else.

**Comment by dannon** [comment]
@galaxybot test this

File: config/job_conf.xml.sample_advanced [hunk 1]
```diff
@@ -205,6 +205,17 @@
                  zero (no execution) and the stderr/stdout of the k8s job is reported in galaxy (and the galaxy job set
                  to failed) -->
         </plugin>
+        <plugin id="godocker" type="runner" load="galaxy.jobs.runners.godocker:GodockerJobRunner">
```
Comment (jmchilton) [inline config/job_conf.xml.sample_advanced:208]: Can you put a line in here about what Godocker is and maybe a link to the platform.

File: lib/galaxy/jobs/runners/godocker.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+        self.noCert = noCert
+
+    def setToken(self, token):
+        self.token = token
+
+    def http_post_request(self, query, data, header):
+        """ post request with query """
+
+        # remove warnings if using --no-certificate
+        requests.packages.urllib3.disable_warnings()
```
Comment (jmchilton) [inline lib/galaxy/jobs/runners/godocker.py]: Can you please remove this line?

**Comment by jmchilton** [comment]
Thanks for the tweaks @dannon and thanks for the job runner @osallou!

**Pull request merged**